DOC: Updated the DataFrame.assign docstring

aeltanawy opened the pull request:
Updated the DataFrame.assign docstring example to use np.arange instead of np.random.randn to pass the validation test.

jschendel reviewed (commented):
Thanks, lgtm aside from one small comment

jschendel commented on pandas/core/frame.py:
```diff
@@ -3353,38 +3353,39 @@ def assign(self, **kwargs):
         --------
-        >>> df = pd.DataFrame({'A': range(1, 11), 'B': np.random.randn(10)})
+        >>> df = pd.DataFrame({'A': range(1, 11),
+        ...                    'B':np.arange(-1.0, 2.0, 0.3)})
```
needs a space between `'B':` and `np.arange`

aeltanawy replied:
Thanks @jschendel for spotting this! I'll send a fix right away.

datapythonista commented:
Thanks for the fix, that's much better than using random.

What would you think about simplifying the example even further? I don't think we need 10 rows to show what assign does. And using a "simpler" function than log (e.g. `lambda x: x ** 2`) would probably help users understand the example faster.

aeltanawy commented:
Thanks @datapythonista and @jorisvandenbossche! I'll shorten the table(s) to hold only 2 rows.

jreback commented:
lgtm. @aeltanawy normally don't put 2 commits like this together (e.g. the statsmodels change is already in master)

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3353,38 +3353,23 @@ def assign(self, **kwargs):
         --------
-        >>> df = pd.DataFrame({'A': range(1, 11), 'B': np.random.randn(10)})
+        >>> df = pd.DataFrame({'A': range(1, 3),
+        ...                    'B': np.arange(-1.0, 2.0, 1.5)})
```
Now that there are only few items I think using something like `[1, 2, 3]` would be clearer than `range(1, 3)`. And same for the other parameter.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3353,38 +3353,23 @@ def assign(self, **kwargs):
-        9  10 -0.758542  2.302585
+            A    B      ln_A
+         0  1 -1.0  0.000000
+         1  2  0.5  0.693147
```
I agree with @jorisvandenbossche that not using `lambda` would be better. But I don't think that's possible, given that the column from the dataframe need to be specified. So, unless there is an easy way to avoid lambda, I'd prefer a function where the user can quickly compute the result mentally, and understand faster what happened in the operation. And if it's a real example, even better, I think that makes things even simpler for the reader.

Not sure if this is too complex, but something like that would be better in my opinion:

```
import pandas as pd

df = pd.DataFrame([('liquid', 100.),
                   ('liquid', 356.73),
                   ('gas', -252.87)],
                  index=['water', 'mercury', 'hydrogen'],
                  columns=('state', 'boiling_point_c'))

df.assign(state=lambda x: pd.Categorical(x.state),
          boiling_point_f=lambda x: x.boiling_point_c * 9 / 5 + 32)
```

jorisvandenbossche commented:
> I agree with @jorisvandenbossche that not using lambda would be better.

Ah, yes, missed that the `np.log` example was also using `lambda` .. Indeed not easy to avoid in this case.

aeltanawy commented:
> @aeltanawy normally don't put 2 commits like this together (e.g. the statsmodels change is already in master)

Ah! Sorry for that, I messed up my branch when fetching upstream then rebasing. Hopefully all sorted out now.

@datapythonista inspired by your example, how about using simply temperatures in Celsius and Fahrenheit indexed with cities?

datapythonista commented:
Sure, I just gave an example to illustrate my point, and not leave an abstract comment. But feel free to use the example that you find more useful/clear for users.

aeltanawy commented:
I apologize for adding already committed commits, again! Anyone willing to show me what I'm doing wrong? (please let me know if this is not the place to ask)

My workflow is as follows:
1. edit the script in my local branch
2. git add, the git commit
3. then git pull origin <local branch>
4. then git push origin <local branch>

jorisvandenbossche commented:
@aeltanawy no problem! 

Looking at your workflow, the one thing that is normally not needed is third step (`git pull origin doc`).

Now, I am not really sure how doing that ended up in having those extra commits, but let's try to solve it.

I would first try:

```
git pull upstream master
git push origin doc
```

and check if that fixes the commits here in the PR (the above updates the PR with changes that have happened in the meantime in the upstream (the pandas-dev/pandas repo) master branch). 

If that didn't fix it, another approach is doing a rebase:

```
git fetch upstream
git rebase upstream/master
git push origin doc --force
```

(note here you will need to force push because rebasing means rewriting the history of the branch, not simply adding commits to the branch as merging does).

If you have more questions, feel free to ask!

aeltanawy commented:
I see now what I was doing wrong. I should have amend my commit instead of creating new ones, and not to fetch upstream and merge origin master. 

@jschendel, shall I close this PR and start a new one with the new edits?

datapythonista committed:
Working on the assign docstring

datapythonista commented:
@aeltanawy I fix the git problems for you. If you simply do a `git pull` in your branch, you should be able to continue working with the docstring.

aeltanawy committed:
DOC: cont'd simplified examples in DataFrame.assign docstring

jorisvandenbossche reviewed (approved):
Looks very nice now!

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3354,38 +3354,23 @@ def assign(self, **kwargs):
         --------
-        >>> df = pd.DataFrame({'A': range(1, 11), 'B': np.random.randn(10)})
+        >>> df = pd.DataFrame({'temp_c': (17.0, 25.0)},
+                                index=['portland', 'berkeley'])
```
Do you mind displaying `df` content after creating it? I think it'll help users understand faster that a new column is created in the assign later on.

Also, do you think it could look better capitalizing the names of the cities?

aeltanawy replied:
I second your thoughts! Will send an update soon.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3354,38 +3354,23 @@ def assign(self, **kwargs):
+        >>> df.assign(temp_f=newcol)
+                  temp_c  temp_f
+        portland    17.0    62.6
+        berkeley    25.0    77.0
```
May be we could manually create a `Series` with Kelvin degrees and assign it. I personally find it a bit confusing that both examples do the same thing, but in a different way. Or we should say it if you think this is better.

I think a single example:
```
>>> df.assign(temp_f=lambda...
>>>                 temp_k=kelvin_series)
```

could illustrate both cases in a simpler way.

jschendel replied:
Maybe expanding the exposition could clear this up a bit?  E.g, "Where the value already exists and is inserted:" --> "Alternatively, the same behavior can be achieved by directly referencing an existing Series or list-like:". 

Would be fine with something along the lines of your suggestion too though.

aeltanawy replied:
I'm leaning more towards keeping this example but to change its exposition to: "Alternatively, the same behavior can be achieved by directly referencing an existing Series or list-like".

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3354,38 +3354,23 @@ def assign(self, **kwargs):
+        portland    17.0    62.6
+        berkeley    25.0    77.0
 
         Where the keyword arguments depend on each other
```
I think this last example is not useful anymore, or do you think it is?

jschendel replied:
This example is still necessary in some form, as it shows a non-obvious Python 3.6+ feature that was recently added (#18852).  It could definitely be improved though.

Reversing the setup and starting from a DataFrame with Fahrenheit data:
```python
In [2]: df = pd.DataFrame({'temp_f': [-40, 0, 32, 100, 212]})

In [3]: df
Out[3]:
   temp_f
0     -40
1       0
2      32
3     100
4     212
```

Then what this is section is trying to illustrate is that, in Python 3.6+, you can create multiple columns within the same `assign` where one of the columns depends on another one defined within the same `assign`.

For example, if you wanted to add both Celsius and Kelvin columns, you could use a formula to build Celsius from Fahrenheit, then use the Celsius column to create the Kelvin one (instead of using a more complex formula referencing Fahrenheit), all from within the same `assign`:
```python
In [4]: df.assign(temp_c=lambda df: 5 / 9 * (df['temp_f'] - 32), 
                  temp_k=lambda df: df['temp_c'] + 273.15)
Out[4]:
   temp_f      temp_c      temp_k
0     -40  -40.000000  233.150000
1       0  -17.777778  255.372222
2      32    0.000000  273.150000
3     100   37.777778  310.927778
4     212  100.000000  373.150000
```

There have also been some more generic questions about how to add multiple columns using `assign` (e.g. [StackOverflow](https://stackoverflow.com/q/42101382/3339965)).  It might be beneficial to first give an example similar to the one above, but where you create Kelvin from Fahrenheit, to simply show how to add multiple columns within the same `assign`.  Then the 3.6+ behavior could be discussed using something similar to my example above.

aeltanawy replied:
Ah! I understand now why this example is needed. Let me add a different one that better portrays its usage.

datapythonista reviewed (commented):
Looks quite good, just couple of ideas to make the examples simpler.

jreback commented:
@datapythonista merge when satisfied

aeltanawy committed:
DOC: adjusted docstring examples in DataFrame.assign to illustrate python3.6+ feature.

aeltanawy commented:
Here are the recent changes to the DataFrame.assign docstring:

1. Showed the output of the initial dataframe.
2. Re-wrote the third example to illustrate python3.6+ feature of being able to assign multiple columns that depends on each other. The example is derived from @datapythonista and @jschendel  comments.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3250,48 +3250,34 @@ def assign(self, **kwargs):
         Examples
         --------
-        >>> df = pd.DataFrame({'A': range(1, 11), 'B': np.random.randn(10)})
+        >>> df = pd.DataFrame({'temp_c': (17.0, 25.0)},
```
Can you use a list instead of a tuple for the data? I think it's more conventional.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3250,48 +3250,34 @@ def assign(self, **kwargs):
         --------
-        >>> df = pd.DataFrame({'A': range(1, 11), 'B': np.random.randn(10)})
+        >>> df = pd.DataFrame({'temp_c': (17.0, 25.0)},
+                                index=['Portland', 'Berkeley'])
```
This is missing `...` at the same level of the `>>>`, and the indentation is not correct

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3250,48 +3250,34 @@ def assign(self, **kwargs):
+        Alternatively, the same behavior can be achieved by directly
+        referencing an existing Series or list-like:
+        >>> newcol = df['temp_c'] * 9 / 5 + 32
+        >>> df.assign(temp_f=newcol)
```
I'd use the expression directly in the assignment instead (i.e. `>>> df.assign(temp_f=df['temp_c'] * 9 / 5 + 32)`),  but it's your call

aeltanawy replied:
From my understanding, this example is to show that you can refer to an existing list or series and assign it to the df. The example above it is already using direct assignment. Do you think it is not necessary to show this?

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3250,48 +3250,34 @@ def assign(self, **kwargs):
+        where one of the columns depends on another one defined within the same
+        assign:
+        >>> df.assign(temp_f=lambda x: x['temp_c'] * 9 / 5 + 32,
+                        temp_k=lambda x: (x['temp_f'] +  459.67) * 5 / 9)
```
same as before regarding `...` and indentation

datapythonista reviewed (changes requested):
Thanks for the changes, the example looks great, except for the formatting issues.

Can you run `./scripts/validate_docstrings.py pandas.DataFrame.assign` after you perform the requested changes? that should tell you if all the formatting problems I commented are fixed.

discort committed:
Fixed loffset with numpy timedelta (#22482)

gfyoung committed:
CLN: Rename 'n' to 'repeats' in .repeat methods (#22574)

For Index and MultiIndex.

xref gh-14645.

elmq0022 committed:
DOC: Updating DataFrame.merge docstring (#22141)

WillAyd committed:
TST: Add capture_stderr decorator to test_validate_docstrings (#22543)

datapythonista reviewed (commented):
Besides the previous comments, can you remove the `df` in the Returns section (just leave DataFrame in the first line).

And also, add the `**` to the `kwargs` parameter, and change the type, so the first line is `**kwargs : dict of {str: callable or Series}`

Thanks!

gfyoung committed:
BLD: Fix openpyxl to 2.5.5 (#22601)

2.5.5 --> 2.5.6 broke compatibility
with pandas Timestamp objects.

Closes gh-22595.

jbrockmendel committed:
Use dispatch_to_series where possible (#22572)

discort committed:
BUG: resample with TimedeltaIndex, fenceposts are off (#22488)

CAM-Gerlach committed:
DOC: Update link and description of the Spyder IDE in Ecosystem docs (#22136)

seantchan committed:
DOC: Improve the docstring of DataFrame.equals() (#22539)

h-vetinari committed:
TST: fixturize series/test_alter_axes.py (#22526)

jorisvandenbossche committed:
TST: restructure internal extension arrays tests (split between /arrays and /extension) (#22026)

rhysparry committed:
TST: Fix skipping test due to lack of connectivity (#22598)

`.format()` was expecting keyword arguments. Updated to match other skips nearby.

mroeschke committed:
API: Add CalendarDay ('CD') offset (#22288)

HyunTruth committed:
CLN/DEPR: removed deprecated as_indexer arg from str.match() (#22626)

gfyoung committed:
BUG: NaN should have pct rank of NaN (#22600)

Closes gh-22519.

alimcmaster1 committed:
Set hypothesis healthcheck (#22597)

jbrockmendel committed:
Implement delegate_names to allow decorating delegated attributes (#22599)

jbrockmendel committed:
[PERF] use numexpr in dispatch_to_series (#22284)

jbrockmendel committed:
Fix incorrect DTI/TDI indexing; warn before dropping tzinfo (#22549)

jbrockmendel committed:
[CLN] More cython cleanups, with bonus type annotations (#22283)

jbrockmendel committed:
move rename functionality out of internals (#21924)

jbrockmendel committed:
TST: Continue collecting arithmetic tests (#22559)

sideeye committed:
BUG: fix failing DataFrame.loc when indexing with an IntervalIndex (#22576)

Peque committed:
DOC: Update `month_name` and `day_name` docstrings (#22544)

h-vetinari committed:
CLN: tests for str.cat (#22575)

Moisan committed:
DOC: Fix to_latex docstring. (#22516)

aeltanawy committed:
DOC: Adjusted DataFrame.assign docstring

pep8speaks commented:
Hello @aeltanawy! Thanks for updating the PR.

 - There are no PEP8 issues in the file [`pandas/core/frame.py`](https://github.com/aeltanawy/pandas/blob/1fa9bc5983808d6d62a4625a49af7c8ee1f0079d/pandas/core/frame.py) !

aeltanawy commented:
@datapythonista, I have adjusted DataFrame.assign docstring with your comments. I've also ran ./scripts/validate_docstrings.py pandas.DataFrame.assign till I got 'the Docstring for "pandas.DataFrame.assign" correct. :)'.

Let me know your thoughts on this commit. Thanks!

simonjayhawkins committed:
TST: add test to io/formats/test_to_html.py to close GH6131 (#22588)

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3220,7 +3220,7 @@ def assign(self, **kwargs):
         Parameters
         ----------
-        kwargs : keyword, value pairs
+        **kwargs : dict of {str: callable or series}
```
can you use `Series` (with capital `S`) please?

datapythonista commented on pandas/core/frame.py:
```diff
@@ -3250,48 +3250,34 @@ def assign(self, **kwargs):
+        Berkeley    25.0    77.0
+
+        Alternatively, the same behavior can be achieved by directly
+        referencing an existing Series or list-like:
```
IMO `sequence` would be more accurate than `list-like`. A tuple is a sequence, and could be used, but I wouldn't say it's a list-like (it doesn't implement the API of list).

datapythonista reviewed (commented):
lgtm, just two details that if you can change them, it'd be great.

jschendel committed:
DOC/CLN: small whatsnew fixes (#22659)

aeltanawy committed:
DOC: adjusted the grammer in DataFrame.assign docstring.

aeltanawy commented:
Thanks @datapythonista! Adjusted `series` and `list-like` to `Series` and `sequence` respectively.

topper-123 committed:
DOC: Add cross references to advanced.rst (#22671)

datapythonista reviewed (approved):
lgtm, thanks a lot for the work in this @aeltanawy

matthewgilbert committed:
DOC: Add section on MultiIndex.to_frame() ordering (#22674)

jbrockmendel committed:
TST: Avoid DeprecationWarnings (#22646)

jbrockmendel committed:
TST: Collect/Use arithmetic test fixtures (#22645)

jbrockmendel committed:
pythonize cython code (#22638)

TomAugspurger committed:
API: register_extension_dtype class decorator (#22666)

TomAugspurger committed:
TST: Close ZipFile in compression test (#22679)

* Updates HypothesisCheck setting
* Skips tests for old xlrd

gfyoung committed:
CLN: Standardize searchsorted signatures (#22670)

The parameter is "value" across the board.

xref gh-14645.

TomAugspurger committed:
DEPR: Removed styler shim (#22691)

* DEPR: Removed styler shim

* sphinx

rth committed:
TST Use pytest.raises instead of legacy constructs (#22681)

alimcmaster1 committed:
Fix test_sql pytest fixture warnings (#22515)

* Avoid calling pytest fixtures directly

henriqueribeiro committed:
API: Add 'name' as argument for index 'to_frame' method (#22580)

reidy-p committed:
BUG: Incorrect addition of Week(weekday=6) to DatetimeIndex (#22695)

* BUG: Incorrect addition of Week(weekday=6) to DatetimeIndex

h-vetinari committed:
ASV: more for str.cat (#22652)

jbrockmendel committed:
TST: Test for bug fixed during #22534 discussion (#22694)

ratijas committed:
Fix broken link in install.rst (#22716)

troels committed:
BUG: Make sure that sas7bdat parsers memory is initialized to 0 (#21616) (#22651)

AaronCritchley committed:
API:  Make .shift always copy (Fixes #22397) (#22517)

nmusolino committed:
TST: Add test of DataFrame.xs() with duplicates (#13719) (#22294)

gfyoung committed:
DEPR: Standardize searchsorted signature (#22672)

"value" is the law of the land.

xref gh-14645.

Follow-up to gh-15601.

h-vetinari committed:
TST/CLN: break up & parametrize tests for df.set_index (#22236)

TomAugspurger committed:
TST: Mock clipboard IO (#22715)

* Attempt to fix clipboard tests

* note

* update

* update

* doc

SandrineP committed:
removing superfluous reference to axis in Series.reorder_levels docstring (#22734)

aeltanawy commented:
Thanks a lot @datapythonista! I've learned a lot!

Is there is any next steps I should do for this case, such as closing this pull request?

mroeschke committed:
CLN/DOC: Refactor timeseries.rst intro and overview (#22728)

* CLN/DOC: Refactor timeseries.rst intro and overview

* Address review

* Forgot missing is

datapythonista commented:
can you merge master into your branch and update the PR please? Seems like some checks failed, and this way they'll rerun and we can merge the changes. Other than that, it's in our side.

In case it's not clear what I meant, in your branch do:
`git fetch upstream ; git merge upstream/master ; git push`

mroeschke committed:
CLN: Remove unused imports in pyx files (#22739)

vitoriahmc committed:
CLN: Removes module pandas.json (#22737)

simonjayhawkins committed:
TST/CLN: remove duplicate data file used in tests (unicode_series.csv) (#22723)

troels committed:
BUG: Some sas7bdat files with many columns are not parseable by read_sas (#22628)

topper-123 committed:
DOC: improve doc string for .aggregate and .transform (#22641)

HannahFerch committed:
BUG: DataFrame.apply not adding a frequency if freq=None (#22150) (#22561)

tswast committed:
[ENH] pull in warning for dialect change from pandas-gbq. (#22557)

JesperDramsch committed:
DOC: Updating str_repeat docstring (#22571)

jbrockmendel committed:
use fused types for reshape (#22454)

jbrockmendel committed:
use fused types for parts of algos_common_helper (#22452)

lucadonini96 committed:
DOC: Updating the docstring of Series.str.extractall (#22565)

realead committed:
BUG: don't mangle NaN-float-values and pd.NaT (GH 22295)  (#22296)

newinh committed:
DOC: Expose ExcelWriter as part of the Generated API (#22359)

Moisan committed:
Test in scripts/validate_docstrings.py that the short summary is always one line long (#22617)

illegalnumbers committed:
fix raise of TypeError when subtracting timedelta array (#22054)

closes #21980

makbigc committed:
Bug: Logical operator of Series with Index (#22092) (#22293)

* Fix bug #GH22092

* Update v0.24.0.txt

* Update v0.24.0.txt

* Update ops.py

* Update test_operators.py

* Update v0.24.0.txt

* Update test_operators.py

Moisan committed:
DOC: Fix Series nsmallest and nlargest docstring/doctests (#22731)

h-vetinari committed:
Fixturize tests/frame/test_api and tests/sparse/frame/test_frame (#22738)

troels committed:
BUG SeriesGroupBy.mean() overflowed on some integer array (#22653)

TomAugspurger committed:
TST: Fail on warning (#22699)

JayOfferdahl committed:
BUG: Allow IOErrors when attempting to retrieve default client encoding. (#21531)

alimcmaster1 committed:
API: Git version (#22745)

topper-123 committed:
DOC: add more links to the API in advanced.rst (#22746)

Moisan committed:
DOC: Fix DataFrame.to_xarray doctests and allow the CI to run it. (#22673)

azure-pipelines[bot] committed:
Set up CI with Azure Pipelines (#22760)

TomAugspurger committed:
CI: Fix travis CI (#22765)

TomAugspurger committed:
CI: Publish test summary (#22770)

yeojin-dev committed:
BUG: Check types in Index.__contains__ (#22085) (#22602)

aeltanawy committed:
Merge remote-tracking branch 'upstream/master' into doc

aeltanawy commented:
Do I need to rebase my changes against master?

datapythonista commented:
It's like the other way round I'd say. You get the latest changes from pandas `git fetch master`, you merge those changes into your branch `git merge upstream/master`, so your branch has your changes on top of the latest pandas, not the pandas of the time you started the changes, and then you update the PR with them `git push`. Besides having your changes on the latest development version of pandas, any update to the PR will make the continuous integration to run again. So the checks that failed (unrelated to your changes), will luckily pass this time.

aeltanawy committed:
Merge remote-tracking branch 'upstream/master' into doc

TomAugspurger commented:
Git changes look OK.

@aeltanawy could you remove `-assign` from https://github.com/pandas-dev/pandas/blob/0480f4c183a95712cb8ceaf5682c5b8dd02e0f21/ci/doctests.sh#L24? Then the docstring will be tested on each PR.

aeltanawy commented:
Looks like all checks passed.

@TomAugspurger, I'll create another pull request for removing -assign so as not to confuse this one which is entirely a DOC case. How does this sound?

datapythonista commented:
@aeltanawy it's better if you do it in this one, as removing the `-assign` will make the CI validate that your examples pass the doctests already.

aeltanawy committed:
Removing -assign from pandas/ci/doctests.sh

aeltanawy commented:
ah! 
removed -assign from pandas/ci/doctests.sh

datapythonista commented:
Thanks @aeltanawy, merging on green.

jschendel commented:
Thanks @aeltanawy!

jorisvandenbossche commented:
@aeltanawy Thanks a lot!
